src/wallet/wallet.cpp

```diff
@@ -4876,125 +4890,118 @@ void CWallet::GetFilteredNotes(
-                        hSig,
-                        (unsigned char) j);
+                ++itr_sprout;
+
+                if (FilterCommon(wtx) || FilterSprout(wtx, nd, paymentaddress, jsop))
+                    continue;
+
+                const auto decrypted = wtx.DecryptSproutNote(jsop).first;
 
                 sproutEntries.push_back(SproutNoteEntry {
-                    jsop, pa, plaintext.note(pa), plaintext.memo(), wtx.GetDepthInMainChain() });
-
-            } catch (const note_decryption_failed &err) {
-                // Couldn't decrypt with this spending key
-                throw std::runtime_error(strprintf("Could not decrypt note for payment address %s", EncodePaymentAddress(pa)));
-            } catch (const std::exception &exc) {
-                // Unexpected failure
-                throw std::runtime_error(strprintf("Error while decrypting note for payment address %s: %s", EncodePaymentAddress(pa), exc.what()));
+                    jsop, paymentaddress, decrypted.note(paymentaddress), decrypted.memo(), wtx.GetDepthInMainChain(), ts
+                });
             }
         }
+        else if (boost::get<libzcash::SaplingPaymentAddress>(&zaddr) != nullptr) {
+            SaplingPaymentAddress paymentaddress = *boost::get<libzcash::SaplingPaymentAddress>(&zaddr);
+            paymentaddress_hash = paymentaddress.GetHash();
+            auto itr_sapling = index.lower_bound(std::make_tuple(NoteType::sapling, paymentaddress_hash, *timestamp));
 
-        for (auto & pair : wtx.mapSaplingNoteData) {
-            SaplingOutPoint op = pair.first;
-            SaplingNoteData nd = pair.second;
+            while (itr_sapling != index.end() && itr_sapling->address == paymentaddress_hash) {
 
-            auto maybe_pt = SaplingNotePlaintext::decrypt(
-                wtx.vShieldedOutput[op.n].encCiphertext,
-                nd.ivk,
-                wtx.vShieldedOutput[op.n].ephemeralKey,
-                wtx.vShieldedOutput[op.n].cmu);
-            assert(static_cast<bool>(maybe_pt));
-            auto notePt = maybe_pt.get();
+                auto nd = pwalletMain->mapWallet[itr_sapling->hash].mapSaplingNoteData[*itr_sapling->op];
+                auto wtx = mapWallet[itr_sapling->hash];
 
-            auto maybe_pa = nd.ivk.address(notePt.d);
-            assert(static_cast<bool>(maybe_pa));
-            auto pa = maybe_pa.get();
+                SaplingOutPoint op = itr_sapling->op.value();
+                auto ts = itr_sapling->timestamp;
 
-            // skip notes which belong to a different payment address in the wallet
-            if (!(filterAddresses.empty() || filterAddresses.count(pa))) {
-                continue;
-            }
+                ++itr_sapling;
 
-            if (ignoreSpent && nd.nullifier && IsSaplingSpent(*nd.nullifier)) {
-                continue;
-            }
+                if (FilterCommon(wtx) || FilterSapling(wtx, nd, paymentaddress, op))
```

Comment:
It looks to me that you are performing this query in a slightly inefficient way with regards to pagination.

It looks like when getting a query with specified limit (ex. limit=10) you would first call `GetFilteredNotes` and populate two vectors `saplingEntries` and `sproutEntries`. Than, on top of this you apply `getPaginatedEntries` to get the slice of the vector.

It would be better to refactor further and split this method to produce sprout and sapling method seperately.
These new methods could get `offset` and `limit` optional parameters.
You initialize with `countAlreadySkipped = 0`, and skip as many notes as specified by `offset`. Than you start decrypting notes and only decrypt at most `limit` of them. If you've reached the limit you stop the iterator.

This way you would only decrypt at most `limit` of notes. 
Such query mechanism will scale it's complexity linearly with `limit` parameter. 
Current solution scales with number of notes per address, which is better than what we're starting from, but not exactly there yet.

Please let me know what you think

Reply:
Sounds good, giving it a try.